File: frontend/src/types/order.ts
export type CheckoutItemPayload = {
  variantId: number
  productName: string
  sku: string
  attributes: Record<string, unknown>
  unitPrice: number
  quantity: number
}

export type CheckoutRequest = {
  customerName: string
  customerPhone: string
  region: string
  items: CheckoutItemPayload[]
}

export type CheckoutResponse = {
  orderId: string
  /** short number the customer can quote when we call ("№1042") */
  orderNumber: number
  itemsTotal: number
  deliveryFee: number
  grandTotal: number
}

export type OrderItemDto = {
  variantId: number
  productName: string
  sku: string
  attributes: Record<string, unknown>
  unitPrice: number
  quantity: number
}

export type OrderStatus = 'NEW' | 'CONFIRMED' | 'PURCHASED' | 'SHIPPED' | 'DELIVERED' | 'CANCELLED'

/** Admin edit of an order; only the fields sent are changed (empty adminNote clears it). */
export type OrderUpdatePayload = {
  status?: OrderStatus
  weightFee?: number
  adminNote?: string
}

export type OrderDto = {
  id: string
  number: number
  customerName: string
  customerPhone: string
  region: string
  itemsTotal: number
  deliveryFee: number
  grandTotal: number
  createdAt: string
  status: OrderStatus
  statusUpdatedAt: string | null
  /** parcel-weight surcharge agreed at confirmation; null until known */
  weightFee: number | null
  /** grandTotal + weightFee; null until the weight fee is known */
  finalTotal: number | null
  adminNote: string | null
  items: OrderItemDto[]
}

File: order-service/src/main/java/com/adikabuyer/order/domain/Order.java
package com.adikabuyer.order.domain;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OrderBy;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "customer_order")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Order {

    @Id
    private String id;

    /** Short, sequential number customers and admins actually say out loud ("№1042"). */
    @Column(nullable = false, unique = true, updatable = false)
    private Long number;

    @Column(name = "customer_name", nullable = false)
    private String customerName;

    @Column(name = "customer_phone", nullable = false)
    private String customerPhone;

    @Column(nullable = false)
    private String region;

    @Column(name = "items_total", nullable = false)
    private BigDecimal itemsTotal;

    @Column(name = "delivery_fee", nullable = false)
    private BigDecimal deliveryFee;

    @Column(name = "grand_total", nullable = false)
    private BigDecimal grandTotal;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private OrderStatus status = OrderStatus.NEW;

    @Column(name = "status_updated_at")
    private Instant statusUpdatedAt;

    /** Parcel-weight surcharge agreed at confirmation; null until known. */
    @Column(name = "weight_fee")
    private BigDecimal weightFee;

    @Column(name = "admin_note", length = 1000)
    private String adminNote;

    /** What the customer actually pays: the checkout total plus the weight fee once agreed. */
    public BigDecimal getFinalTotal() {
        return weightFee == null ? null : grandTotal.add(weightFee);
    }

    @Builder.Default
    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    @OrderBy("id ASC")
    private List<OrderItem> items = new ArrayList<>();
}
